staging/kos/pkg/controllers/connectionagent/connectionagent.go

```diff
@@ -0,0 +1,1467 @@
+		fabricLatencyHistograms:              fabricLatencyHistograms,
+		attachmentCreateToStatusHistogram:    attachmentCreateToStatusHistogram,
+		attachmentStatusHistograms:           attachmentStatusHistograms,
+		localAttachmentsGauge:                localAttachmentsGauge,
+		remoteAttachmentsGauge:               remoteAttachmentsGauge,
+		attachmentExecDurationHistograms:     attachmentExecDurationHistograms,
+		attachmentExecStatusCounts:           attachmentExecStatusCounts,
+	}
+}
+
+// Run activates the ConnectionAgent: the local attachments informer is started,
+// pre-existing network interfaces on the node are synced, and the worker
+// goroutines are started. Close stopCh to stop the ConnectionAgent.
+func (ca *ConnectionAgent) Run(stopCh <-chan struct{}) error {
+	defer k8sutilruntime.HandleCrash()
+	defer ca.queue.ShutDown()
+
+	// Serve Prometheus metrics
+	http.Handle("/metrics", promhttp.Handler())
+	go func() {
+		klog.Errorf("In-process HTTP server crashed: %s", http.ListenAndServe(MetricsAddr, nil).Error())
+	}()
+
+	ca.stopCh = stopCh
+	ca.initLocalAttsInformerAndLister()
+	go ca.localAttsInformer.Run(stopCh)
+	klog.V(2).Infoln("Local NetworkAttachments informer started")
+
+	if err := ca.waitForLocalAttsCacheSync(stopCh); err != nil {
+		return err
+	}
+	klog.V(2).Infoln("Local NetworkAttachments cache synced")
+
+	if err := ca.syncPreExistingIfcs(); err != nil {
+		return err
+	}
+	klog.V(2).Infoln("Pre-existing interfaces synced")
+
+	for i := 0; i < ca.workers; i++ {
+		go k8swait.Until(ca.processQueue, time.Second, stopCh)
+	}
+	klog.V(2).Infof("Launched %d workers", ca.workers)
+
+	<-stopCh
+	return nil
+}
+
+func (ca *ConnectionAgent) initLocalAttsInformerAndLister() {
+	localAttWithAnIPSelector := ca.localAttWithAnIPSelector()
+
+	ca.localAttsInformer, ca.localAttsLister = v1a1AttsCustomInformerAndLister(ca.kcs,
+		resyncPeriod,
+		fromFieldsSelectorToTweakListOptionsFunc(localAttWithAnIPSelector.String()))
+
+	ca.localAttsInformer.AddIndexers(map[string]k8scache.IndexFunc{attVNIAndIPIndexerName: attVNIAndIPIndexer})
+
+	ca.localAttsInformer.AddEventHandler(k8scache.ResourceEventHandlerFuncs{
+		AddFunc:    ca.onLocalAttAdded,
+		UpdateFunc: ca.onLocalAttUpdated,
+		DeleteFunc: ca.onLocalAttRemoved,
+	})
+}
+
+func (ca *ConnectionAgent) onLocalAttAdded(obj interface{}) {
+	att := obj.(*netv1a1.NetworkAttachment)
+	klog.V(5).Infof("Local NetworkAttachments cache: notified of addition of %#+v", att)
+	ca.queue.Add(parse.AttNSN(att))
+}
+
+func (ca *ConnectionAgent) onLocalAttUpdated(oldObj, obj interface{}) {
+	oldAtt, att := oldObj.(*netv1a1.NetworkAttachment), obj.(*netv1a1.NetworkAttachment)
+	klog.V(5).Infof("Local NetworkAttachments cache: notified of update from %#+v to %#+v", oldAtt, att)
+	if oldAtt.Status.IPv4 != att.Status.IPv4 || oldAtt.Status.AddressVNI != att.Status.AddressVNI {
+		// The only fields affecting local network interfaces handling that can
```

Comment:
If the object is deleted and replaced can't we get an arbitrary change here?

Reply:
No, because the Informer this notification handler is registered to has a [field selector](https://github.com/MikeSpreitzer/kube-examples/blob/discuss-ca-simplification/staging/kos/pkg/controllers/connectionagent/connectionagent.go#L1283) that prevents changes to the attachment node.

Reply:
Yes, I think that should do it.